x-pack/plugins/endpoint/public/lib/action_creator.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+type PayloadType =
+  | []
+  | [unknown]
+  | [unknown, unknown]
+  | [unknown, unknown, unknown]
+  | [unknown, unknown, unknown, unknown]
+  | [unknown, unknown, unknown, unknown, unknown];
+
+/**
+ * Returns a function that will produce a Redux action when called
+ *
+ * @param {String} type
+ */
+export function createActionFactory<Type extends string, Payload extends PayloadType>(type: Type) {
```

Comment:
The Resolver PR i'm working on does not use an action creator. I specify one interface per action and then manually create them. 

Here's how i specify actions:
x-pack/plugins/endpoint/public/embeddables/resolver/store/camera/action.ts

Here's an example of how I create and dispatch them: 
https://github.com/elastic/kibana/pull/53619/files#diff-3fd70733fa0c53321aefd73369b09360R86

second example, from implementation:
https://github.com/elastic/kibana/pull/53619/files#diff-b5915dbc86f5686a7835fc3298e4261aR62

The store, reducers, and dispatch functions are all typed with a single action type, which is the union of all individual action types:
https://github.com/elastic/kibana/pull/53619/files#diff-118fa940615495cf306a955b1341de5bR10

Since I intend to use redux middleware to model side effects, it isn't necessary to have a function that creates the action. 

Type inference works just fine in my editor, and documenting the parts of the action are easy. This shows the 'userZoomed' action, and the documentation for the payload.
![image](https://user-images.githubusercontent.com/35559/71693968-c6eefd80-2d7b-11ea-8b80-d9f324168ed9.png)

So anyway, how would you feel about dropping the action creators? :P

Reply:
@oatkiller I'm ok dropping it. It certainly gets way from defining the ugly PayloadType tuple union (which I have heard @kqualters-elastic call _"modern art"_  😝 🤣 )

A few observations (pros and cons): 
- this approach, of just typing the action out as a interface, is certainly clearer when looking at the `payload` and allows for `payload` values of more than just type `array`
- in comparison to our prior approach, it makes it a little more verbose in that you have to define the dispatch action object manually every time (on the flip side - its clearer what we are actually dispatching)
- we were able to use the action creator `.type` property to reference the Action type during equality checks (like in reducers or sagas) where now we will have to type out the actual action `type` string, This opens up the code for more typos and makes refactoring a bit harder since we would now have to do global search-and-replace (probably minor, since I don't see us doing lots of refactoring).

So... 😃 all that being said - I'm 👍 with dropping this helper file and going with manual typing of all action interfaces.

Thanks for the feedback.

ps. Will hold off on making changing in order to give other an opportunity to comment on this topic

Reply:
I agree with dropping it, that ugly type was just for backwards compatibility.

Reply:
@paul-tavares I agree with the above, we should drop it and stay consistent with the resolver embeddable

Reply:
@paul-tavares About the 'search and replace' issue: The language server for typescript (tsserver) supports 'find references' and that can handling finding a property of an interface. Here is a screenshot where I'm finding all references to an action type: